lib/auth/auth_with_roles.go

```diff
@@ -1313,10 +1313,23 @@ func (a *ServerWithRoles) GetNode(ctx context.Context, namespace, name string) (
 	return node, nil
 }
 
-func (a *ServerWithRoles) checkUnifiedAccess(resource types.ResourceWithLabels, checker resourceAccessChecker, filter services.MatchResourceFilter, resourceAccessMap map[string]error) (bool, error) {
+// resourceAccess is used for different rbac results when filtering items
+type resourceAccess struct {
+	// baseAuthChecker is set when a user's auth context is extended during a
+	// searchAsRoles request
+	baseAuthChecker resourceAccessChecker
+	// kindAccessMap is used to check errors for list/read verbs per kind
+	kindAccessMap map[string]error
+	// paginatedResources are resources that have passed the match check and are to be returned in the request
+	paginatedResources []*proto.PaginatedResource
+}
+
+// checkUnifiedAccess will allow the iterator in the unified cache to continue based on match and will populate
+// the passed in resourceAccess.PaginatedResources slice when matches are found.
+func (a *ServerWithRoles) checkUnifiedAccess(resource types.ResourceWithLabels, checker resourceAccessChecker, filter services.MatchResourceFilter, resourceAccess *resourceAccess) (bool, error) {
```

Comment:
This api smells off to me. Returning a bool, writing results to the parameters, ignoring the result of `(Iterate|Get)UnifiedResources(ByIDs)` at other places, somethings not right. I see what the goal is and why it's tricky though

Maybe we're too tied to the current IterateUnifiedResources impl, if that could be changed I feel like the call site should look roughly more like:
```go
var paginatedResources []*proto.PaginatedResource
err := a.authServer.UnifiedResourceCache.IterateUnifiedResources(ctx, func(resource types.ResourceWithLabels) (error) {
	canAccess, canAccessWithRequest, err := a.checkUnifiedAccess(resource, checkUnifiedAccessParams)
	if err != nil {
		return trace.Wrap(err)
	}
	if canAccess || canAccessWithRequest {
		paginatedResources = append(paginatedResources, services.MakePaginatedResource(resource, canAccessWithRequest))
	}
}
```

Reply:
> Maybe we're too tied to the current IterateUnifiedResources impl

After fiddling a bit, I think I _am_ tied too the current `IterateUnifiedResources` implementation. If we (for example) only returned a `nextKey` and `error` from it, then we are still "ignoring" the results from the internal `getRange` that is returned from the cache.  I'm not against it too much, and won't die on this hill, but it its really just masking the issue closer to the source (which I guess can be argued is better than further down the road). 

I've updated _sorta_ based on @rosstimothy suggestion. But instead, I've added these values to the `resourceAccess` struct and made `checkUnifiedAccess` (-> `resourceAccess.checkAccess`) a receiver method for it.

I'm sorta grasping at straws, and maybe I'm too in the weeds right now, but my goal with this current change was to make it easy to follow. It makes sense to me but I've been staring at it for days so please let me know how it is to read now. https://github.com/gravitational/teleport/pull/40328/commits/182800c30e4fbf36dc0d00d3c131ad9da9d01554

Reply:
I like the new version a lot better